### DISCUSSION TITLE
How do I build my own matrix

### DISCUSSION THREAD

**Discussion by aaaaaaqing** [idaholab/moose#25045] [Q&A General]
I know that moose framework is very excellent, especially its encapsulation is very excellent, but this also bothers me to use it, because for some reasons, I need to build the matrix by myself, but I do not understand the principle of moose matrix construction and the underlying code, so how should I read and understand it

**Comment by GiudGiud** [comment]
@tanoret can comment on this. 
@grmnptr as well

**Reply by aaaaaaqing** [reply]
thank u
So how should I contact them?

**Reply by lindsayad** [reply]
they can respond here

**Comment by lindsayad** [comment]
What matrix do you want to build? If you want to build the matrix from just some components of your residual/Jacobian objects, e.g. mass/stiffness matrices, then you can achieve this using the tagging system

**Reply by aaaaaaqing** [reply]
First of all, thank you for your reply. I want to build the Jacobi matrix and residuals I want by myself, the size of this Jacobi matrix is 3N✕3N (n is the number of nodes), the size of the residuals is 3N, how can I build my own matrix in Moose? I noticed the underlying code in the moose kernel (kernel. C) is implemented in the construction of the matrix, can this be achieved by modifying the underlying code of the moose kernel? Or there are other suggestions.

**Reply by lindsayad** [reply]
If you really want to do your own thing, you can write a custom executioner like @grmnptr is doing [here](https://github.com/idaholab/moose/pull/22699/files#diff-8a3cf4c317d32cf8f4009f4d821c43c3ee46702144f38110292c5e67759ee893)

**Reply by grmnptr** [reply]
Thanks @lindsayad ! Yes, if you really need to manipulate the matrices manually, you can access them through the libmesh systems and set values using something like what I do in the linked PR.

```
void
SIMPLE::relaxMatrix(SparseMatrix<Number> & matrix,
                    const Real relaxation_parameter,
                    NumericVector<Number> & diff_diagonal)
{
for (auto row_i = matrix.row_start(); row_i < matrix.row_stop(); row_i++)
   matrix.set(row_i, row_i, diff_diagonal(row_i));
matrix.close();
}
```

```
NonlinearImplicitSystem & momentum_system =
      dynamic_cast<NonlinearImplicitSystem &>(_momentum_systems[0]->system());
SparseMatrix<Number> & mmat = *(momentum_system.matrix);
NumericVector<Number> & rhs = *(momentum_system.rhs);
```

**Reply by YaqiWang** [reply]
Just to make sure one thing: Deriving your own kernels allows you to have your own residual and jacobian for your equation system. This might be what you want.

**Reply by aaaaaaqing** [reply]
Thanks for the reply from the experts, but I don't fully understand the code you mentioned.First, I need to know the steps the kernel takes to build residuals and Jacobian matrices, especially when there are coupling terms. Second, when there are multiple custom kernels, how does moose integrate them. Finally, if I want to know if I have a Jacobian matrix and a residual vector, then how can I solve this matrix using moose (because moose is excellent for parallel computing and adaptive grid systems)

**Reply by aaaaaaqing** [reply]
If it is convenient to tell me that the code files are constructing matrices (processing multiple custom kernels, processing multiple coupled variables, and assembling multiple matrices into the final matrix), then I can understand the principle by reading the code

**Reply by GiudGiud** [reply]
look at Kernel.C.

kernels build diagonal terms by implementing the computeQpJacobian and off-diagonals by computeQpOffDiagJacobian
kernels are called in an elemental loop, see include/loops/ThreadedElementLoop.h

in kernel.C/.h, we process these terms and use Assembly routines to propagate these contributions.
in Assembly.C/.h, you can look at how the matrix is assembled

this is documented here:
https://mooseframework.inl.gov/source/base/Assembly.html
https://mooseframework.inl.gov/syntax/Kernels/index.html

I think you should consider taking Yaqi s advice and try to implement your problem the MOOSE way with regular kernels and boundary conditions (e.g. not Peter's way, that's a very advanced way).

**Reply by aaaaaaqing** [reply]
Yes, you are right, the method of using moose is much simpler and suitable for me. But for my problem, I don't know whether moose's method can solve it.

The first thing I want to know is whether moose uses full Newtonian coupling to calculate?

Secondly, there is a residual term in the algorithm I use: _grad_test(i)*W(ij)*F(i)+_grad_test(i)*W(ij)*F(j), where _grad_test(i) is the gradient of the scalar interpolation function (gradient of the test function) of the point i, and W(ij) is the vector interpolation function of the edge ij, F(i) and F(j) is an expression about the values that need to be solved at points i and j. I don't know how to implement this in moose, so I always wanted to bypass moose and directly build a matrix and then put it into moose for solving. If it can be implemented in moose, then I think moose will be awesome

**Reply by GiudGiud** [reply]
We can compute full Jacobians in MOOSE. It s the default for Newton's method iirc. 

So i and j are quadrature points here?
Is F analytical? If you had the variable values, can you just compute F directly?

MOOSE handles quadrature points one by one. So this doesnt look like a classic MOOSE kernel because of the mixed indexing. However with a tad more work you can **either**:
- override computeResidual instead of computeQpResidual, and have such expressions
- write custom computeQpResidual that indexes the other quantities at other quadrature points than the _qp index. They are pre-computed by elements so they are all available

**Reply by aaaaaaqing** [reply]
Are there any examples of this, even if you just use vector interpolation functions to discretize variables

**Reply by aaaaaaqing** [reply]
The problem that bothers me most is how to use both Ni, which is a scalar interpolation function, and Wij, which is a vector interpolation function

**Reply by lindsayad** [reply]
You're now asking the same question you asked over on https://github.com/idaholab/moose/discussions/25046 . Thank you for sharing  an equation you want to model over there. It helps make things more clear. If that is the physics you want to model, then you could use a Galerkin method with shape functions of type `LAGRANGE_VEC` and then you have no need for both vector and scalar basis functions. You only need the vector basis. Do you know what kind of bases you want to use to describe your vector current?

**Reply by aaaaaaqing** [reply]
Thanks for your help, I'll understand what you're talking about LAGRANGE_VEC shape functions
I am clear about the vector basis function I want to use, for example, for rectangular elements, W1=1/4*(1+Xij)*E1, this represents the vector basis function of side 1, where E represents the direction of the basis function, and Xij indicates that the basis function reaches a maximum of 1/2 at edge x=Xij, and a minimum of 0 at x=-Xij

**Reply by lindsayad** [reply]
Is there a literature name for this basis?

**Reply by aaaaaaqing** [reply]
The equation here is not exactly the same as i did in #25046, because i am using the equation _grad_test (i) *W (ij) *F (i) +_grad_test (i) *W (ij) *F (j), where F(I) and F(j) are an expression of the value of the solution at the node, You can think of it as the coefficient of the basis function when the value to be solved is discretized.

It doesn't really matter what the basis function is, as long as it's a general vector basis function.

After I read LAGRANGE_VEC, I learned that it is a variable that contains both shape function and solution, is my understanding correct? If this is the case, then the variable can only contain a basis function (here I will choose it to contain vector basis functions), in which case, how can I add another scalar function to the equation?

**Reply by aaaaaaqing** [reply]
I have come up with a method that uses a variable of type SCALAR. If this variable has no basis function, then it is the function I am looking for. Then I can use the function constructor with the basis function I use, and my problem is solved. If this scheme works, there is only one problem left: to construct the basis function, I need to know 'for a quadrilateral grid, moose's grid after normalization, what does the individual grid look like, for example, tell me the coordinates of the four vertices'?

**Reply by lindsayad** [reply]
> After I read LAGRANGE_VEC, I learned that it is a variable that contains both shape function and solution, is my understanding correct?

No. It is only a basis. Just as for any finite element solution, the approximate solution is given by uh = sum(uj * phij). The LAGRANGE_VEC basis only describes phij

**Reply by lindsayad** [reply]
> moose's grid after normalization

What do you mean by this?

**Reply by lindsayad** [reply]
> tell me the coordinates of the four vertices'?

You can access the physical coordinates of element nodes by doing `_current_elem->node_ref(0)`, `_current_elem->node_ref(1)`, `_current_elem->node_ref(2)`, `_current_elem->node_ref(3)`.

**Reply by aaaaaaqing** [reply]
No. It is only a basis. Just as for any finite element solution, the approximate solution is given by uh = sum(uj * phij). The LAGRANGE_VEC basis only describes phij：
Ok, I see what you mean, I can use family = LAGRANGE_VEC to create a vector variable, and then use _phi to call the vector interpolation function, which is really a good idea. But I have two questions:

First: How should I call the value uj on the node (not the value on the side)?

Second: For kernels that use variables created with LAGRANGE_VEC, what type of function is _test, scalar or vector

**Reply by aaaaaaqing** [reply]
Or do you mean: I can customize a variable by using family = LAGRANGE_VEC, and then this variable is uh (ij) = sum (F (ui, uj) *phi(ij))? Where F(ui,uj) is a function of the values on the two nodes on the edge of ij, phi (ij) is a vector basis function constructed using family = LAGRANGE_VEC, and uh (ij) is an approximation of the current density constructed on the edge of ij.

Is that right? Does moose have examples of custom variables built?

**Reply by aaaaaaqing** [reply]
Is there a literature name for this basis?：
（1st-order） Nédélec basis function

**Reply by lindsayad** [reply]
> First: How should I call the value uj on the node (not the value on the side)?

I would call it a degree of freedom

> Second: For kernels that use variables created with LAGRANGE_VEC, what type of function is _test, scalar or vector

It is a vector

> （1st-order） Nédélec basis function

We have `NEDELEC_ONE` bases in libMesh. You can see examples of their usage in the electromagnetics module that @cticenhour developed
```
(base) lindad@inl610212:~/projects/moose/modules/electromagnetics(devel)$ git grep -In "family.*=.*NEDELEC_ONE"
test/tests/auxkernels/current_density/current_density.i:32:    family = NEDELEC_ONE
test/tests/auxkernels/current_density/em_current_density.i:20:    family = NEDELEC_ONE
test/tests/auxkernels/current_density/em_current_density.i:27:    family = NEDELEC_ONE
test/tests/auxkernels/current_density/error_test.i:21:    family = NEDELEC_ONE
test/tests/auxkernels/current_density/error_test.i:28:    family = NEDELEC_ONE
test/tests/bcs/vector_robin_bc/portbc_waves.i:31:    family = NEDELEC_ONE
test/tests/bcs/vector_robin_bc/portbc_waves.i:35:    family = NEDELEC_ONE
test/tests/benchmarks/dipole_antenna/dipole.i:21:    family = NEDELEC_ONE
test/tests/benchmarks/dipole_antenna/dipole.i:25:    family = NEDELEC_ONE
test/tests/benchmarks/dipole_antenna/dipole_transient.i:15:    family = NEDELEC_ONE
test/tests/benchmarks/dipole_antenna/dipole_transient.i:19:    family = NEDELEC_ONE
test/tests/benchmarks/evanescent_wave/evanescent_wave.i:37:    family = NEDELEC_ONE
test/tests/benchmarks/evanescent_wave/evanescent_wave.i:41:    family = NEDELEC_ONE
test/tests/interfacekernels/electromagnetic_interfaces/combined_default.i:43:    family = NEDELEC_ONE
test/tests/interfacekernels/electromagnetic_interfaces/combined_default.i:48:    family = NEDELEC_ONE
test/tests/interfacekernels/electromagnetic_interfaces/combined_props.i:48:    family = NEDELEC_ONE
test/tests/interfacekernels/electromagnetic_interfaces/combined_props.i:53:    family = NEDELEC_ONE
test/tests/interfacekernels/electromagnetic_interfaces/parallel.i:43:    family = NEDELEC_ONE
test/tests/interfacekernels/electromagnetic_interfaces/parallel.i:48:    family = NEDELEC_ONE
test/tests/interfacekernels/electromagnetic_interfaces/perpendicular.i:42:    family = NEDELEC_ONE
test/tests/interfacekernels/electromagnetic_interfaces/perpendicular.i:47:    family = NEDELEC_ONE
test/tests/kernels/vector_helmholtz/vector_current_source.i:32:    family = NEDELEC_ONE
test/tests/kernels/vector_helmholtz/vector_current_source.i:36:    family = NEDELEC_ONE
test/tests/kernels/vector_helmholtz/vector_kernels.i:18:    family = NEDELEC_ONE
```
